src/rgw/rgw_op.cc

```diff
@@ -3514,3 +3514,197 @@ void RGWHandler::put_op(RGWOp *op)
+        sleep(60);
+    }
+
+    if ((s->err.http_ret != 200) ||
+        (s->err.ret != 0)) {
+        ldout(s->cct, 0) << "DSS ERROR: Copy object failed during rename op."
+                         << " . Return status: " << s->err.ret
+                         << " . Return HTTP code: " << s->err.http_ret
+                         << " . Return Message: " << copy_op->get_request_state()->err.message
+                         << dendl;
+        s->err.ret = -ERR_RENAME_FAILED;
+        return;
+    }
+    ldout(s->cct, 0) << "DSS INFO: Rename op: copy done" << dendl;
+
+    /* Tweek the request for a delete obj operation and perform delete op */
+    new_obj.dss_duplicate(&(s->object));
+    (s->object).dss_duplicate(&orig_object);
+    RGWDeleteObj_ObjStore_S3* del_op = new RGWDeleteObj_ObjStore_S3;
+    delete_rgw_object(del_op);
+    if ((s->err.http_ret != 200) ||
+        (s->err.ret != 0)) {
+        ldout(s->cct, 0) << "DSS ERROR: Delete object failed during rename op."
+                         << " . Return status: " << s->err.ret
+                         << " . Return HTTP code: " << s->err.http_ret
+                         << " . Return Message: " << del_op->get_request_state()->err.message
+                         << dendl;
+
+        /* Revert the copy op */
+        int ret_orig, ret_newobj;
+        ret_orig = check_obj(s->object);
+        ret_newobj = check_obj(new_obj);
+        if (ret_orig < 0) {
+            // Delete failed but we don't have original object!!
+            if (ret_newobj < 0) {
+                // We are in a soup. Data lost. This is not the case we will ever end up in.
+                s->err.ret = -ERR_RENAME_DATA_LOST;
+            } else {
+                // Everything normal
+                if (!store->ctx()->_conf->fault_inj_rename_op_delete_fail) {
+                    s->err.http_ret = 200;
+                    s->err.ret = 0;
+                }
+            }
+        } else {
+            if (ret_newobj >= 0) {
+                // Delete the new object
+                (s->object).dss_duplicate(&new_obj);
+                RGWDeleteObj_ObjStore_S3* del_op = new RGWDeleteObj_ObjStore_S3;
+                delete_rgw_object(del_op);
+                if ((s->err.http_ret != 200) ||
+                    (s->err.ret != 0)) {
+                    s->err.ret = -ERR_RENAME_NEW_OBJ_DEL_FAILED;
+                } else {
+                    // Indicate that there has been a failure
+                    s->err.ret = -ERR_RENAME_FAILED;
+                }
+            } else {
+                // Log major error. Ask user to file a bug. Why didn't copy fail?
+                s->err.ret = -ERR_RENAME_COPY_FAILED;
+            }
+        }
+        return;
+    }
+    ldout(s->cct, 0) << "DSS INFO: Rename op complete"  << dendl;
+    return;
+}
+
+void RGWRenameObj::perform_external_op(RGWOp* bp)
+{
+    // Fault injection if tester wants so
+    if ((store->ctx()->_conf->fault_inj_rename_op_copy_fail) ||
```

Comment:
The fault injection will fail for copy and delete if any of the conf parameter is set. U cannot fail them separately.